### PR TITLE
Added labels to error spans

### PR DESCRIPTION
**Pull request by jonathandturner**
This adds a label to go along with error spans, so that we can show a message with the error.

### CODE REVIEW

File: components/lark-error/src/lib.rs [hunk 1]
```diff
@@ -35,31 +35,45 @@ use std::sync::Arc;
...
+    pub fn into_labeled_spans(self) -> Vec<LabeledSpan> {
         self.0
     }
 }
 
+/// A span with an associated label.
+/// TODO: We may want to merge this with what's available in error
+/// reporting
+#[derive(Clone, Debug, DebugWith, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct LabeledSpan {
```
Comment (nikomatsakis) [inline components/lark-error/src/lib.rs]: Nit: Maybe we can name this "Diagnostic"? Feels like a "nicer word" to me.

File: components/lark-type-check/src/hir_typeck.rs [hunk 2]
```diff
@@ -423,7 +429,7 @@ where
                         EntityData::LangItem(LangItem::Uint) => uint_type,
                         EntityData::Error(_) => self.error_type(),
                         _ => {
-                            self.record_error(expression);
+                            self.record_error("unknown type for expression".into(), expression);
```
Comment (nikomatsakis) [inline components/lark-type-check/src/hir_typeck.rs]: the idea here was that (for now, anyway) operators like `+` etc require integers. So perhaps a better message might be "can't apply `+` to a `String`" or something? (meh, whatever)

File: components/lark-error/src/lib.rs (hunk 1 above)
Comment (jonathandturner) [reply components/lark-error/src/lib.rs]: Sure, can do. I'm guessing it'll grow into that at some point, so earlier is fine with me

File: components/lark-type-check/src/hir_typeck.rs (hunk 2 above)
Comment (jonathandturner) [reply components/lark-type-check/src/hir_typeck.rs]: Sure, will take a look

**Pull request merged**